File: 202010/20201009/q897.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        if not root:
            return None
        result, stack = TreeNode(), [root]
        pre = result
        while len(stack) != 0:
            while root.left:
               stack.append(root.left)
               root = root.left
            node = stack.pop(-1)
            if node.right:
                stack.append(node.right)
                root = node.right
            pre.right = node
            pre.left = None
            pre = pre.right
        return result.right

File: 202104/20210425/q897.java
import java.util.LinkedList;

// *
//   Definition for a binary tree node.
class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;
  TreeNode() {}
  TreeNode(int val) { this.val = val; }
  TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
  }
}
class Solution {
    public TreeNode increasingBST(TreeNode root) {
        if (root == null) {
            return null;
        }
        TreeNode pre = null, ans = null;
        LinkedList<TreeNode> stack = new LinkedList<>();
        stack.add(root);
        while (!stack.isEmpty()) {
            while (null != root.left) {
                stack.addFirst(root.left);
                TreeNode tmp = root;
                root = root.left;
                tmp.left = null;
            }

            TreeNode remove = stack.pop();

            if (pre == null) {
                pre = remove;
                ans = pre;
            } else {
                pre.right = remove;
                pre = pre.right;
            }

            if (remove.right != null) {
                root = remove.right;
                stack.addFirst(root);
            }
        }
        return ans;
    }
}